File: harness/platform_core_contract/execution_receipt.py
"""Pure Platform Core ExecutionReceipt v1 validation."""

from __future__ import annotations

import re
from typing import Any

from .artifact import (_validate_artifact_references,
                       _validate_artifact_relations,
                       _validate_artifact_values)
from .codec import (ContractError, canonical_json, exact_object, lower_token,
                    validate_document_shape, validate_i64,
                    validate_schema_name)
from .constants import (CANONICALIZATION, EVENT_RANGE_FIELDS,
                        DOCUMENT_INVALID, EXECUTION_RECEIPT_FIELDS, EXECUTOR_FIELDS,
                        EXECUTION_RECEIPT_TYPE_PROFILE,
                        MAX_EXECUTION_ELAPSED_MS, MAX_OBSERVED_COUNT,
                        MAX_OBSERVED_QUANTITY, MAX_REASON_CODES,
                        MAX_RECEIPT_ARTIFACTS, MAX_RECEIPT_BYTES,
                        MAX_UNIX_MILLISECONDS, OBSERVED_USAGE_FIELDS,
                        RECEIPT_VERSION, REFERENCE_MISMATCH,
                        RELATION_MISMATCH, STATE_INVALID)
from .identity import validate_typed_id
from .references import (validate_actor_ref, validate_entity_ref,
                         validate_record_ref, validate_scope_references,
                         validate_scope_values)

_TERMINAL_ATTEMPT_STATES = {"interrupted", "completed", "failed", "uncertain"}
_ADAPTER_VERSION = re.compile(r"(0|[1-9][0-9]{0,8})\.(0|[1-9][0-9]{0,8})\.(0|[1-9][0-9]{0,8})")
_RECORD_ROLES = {
    "approval_ref": "forge.control.approval_record",
    "grant_ref": "forge.control.capability_grant",
}


def validate_execution_receipt(value: Any) -> dict[str, Any]:
    """Return one valid supplied execution observation declaration."""
    validate_document_shape(value, EXECUTION_RECEIPT_TYPE_PROFILE,
                            "ExecutionReceipt")
    node = exact_object(value, EXECUTION_RECEIPT_FIELDS, "ExecutionReceipt")
    if (node["canonicalization"] != CANONICALIZATION or
            validate_i64(node["execution_receipt_version"],
                         "execution_receipt_version", 1) != RECEIPT_VERSION):
        raise ContractError("execution receipt version or canonicalization is unsupported")
    validate_typed_id(node["receipt_id"], "rcp", "receipt_id")
    _validate_time_values(node)
    _validate_executor(node["executor"])
    _validate_declaration_values(node)
    _validate_scope_bindings(node)
    _validate_declaration_references(node)
    _validate_terminal_state(node)
    _validate_relations(node)
    canonical_json(node, MAX_RECEIPT_BYTES)
    return node


def _validate_scope_bindings(node: dict[str, Any]) -> None:
    scope = node["scope_ref"]
    validate_scope_references(scope)
    if (scope["project_snapshot_id"] is None or scope["attempt_id"] is None or
            scope["session_id"] is None or scope["turn_id"] is not None or
            scope["action_id"] is not None):
        raise ContractError(
            "execution receipt requires exact session-level snapshot scope",
            REFERENCE_MISMATCH)
    _exact_ref(node["attempt_ref"], "attempt", scope["attempt_id"], "attempt_ref")
    _exact_ref(node["session_ref"], "session", scope["session_id"], "session_ref")
    _exact_ref(node["source_snapshot_ref"], "project_snapshot",
               scope["project_snapshot_id"], "source_snapshot_ref")


def _exact_ref(value: Any, kind: str, identifier: str, label: str) -> None:
    reference = validate_entity_ref(value, label)
    if reference["entity_type"] != kind or reference["entity_id"] != identifier:
        raise ContractError(f"{label} must exactly match scope_ref",
                            REFERENCE_MISMATCH)


def _validate_time_values(node: dict[str, Any]) -> None:
    _unix_ms(node["started_at_unix_ms"], "started_at_unix_ms")
    _unix_ms(node["ended_at_unix_ms"], "ended_at_unix_ms")
    usage = exact_object(node["observed_usage"], OBSERVED_USAGE_FIELDS,
                         "observed_usage")
    _validate_usage(usage)


def _validate_terminal_state(node: dict[str, Any]) -> None:
    terminal = node["terminal_state"]
    if type(terminal) is not str:
        raise ContractError("terminal_state must be text", DOCUMENT_INVALID)
    if terminal not in _TERMINAL_ATTEMPT_STATES:
        raise ContractError(
            f"terminal_state {terminal!r} is not an Attempt terminal state",
            STATE_INVALID)


def _validate_usage(value: dict[str, Any]) -> None:
    bounds = {
        "cost_usd_micros": MAX_OBSERVED_QUANTITY,
        "elapsed_ms": MAX_EXECUTION_ELAPSED_MS,
        "input_tokens": MAX_OBSERVED_QUANTITY,
        "model_calls": MAX_OBSERVED_COUNT,
        "network_bytes": MAX_OBSERVED_QUANTITY,
        "output_bytes": MAX_OBSERVED_QUANTITY,
        "output_tokens": MAX_OBSERVED_QUANTITY,
        "tool_calls": MAX_OBSERVED_COUNT,
    }
    for field, maximum in bounds.items():
        validate_i64(value[field], f"observed_usage.{field}", 0, maximum)


def _validate_executor(value: Any) -> None:
    node = exact_object(value, EXECUTOR_FIELDS, "executor")
    actor = validate_actor_ref(node["actor_ref"])
    if actor["actor_type"] not in {"agent", "service", "system"}:
        raise ContractError("executor actor_type must be agent, service, or system")
    validate_schema_name(node["adapter_id"], "executor.adapter_id")
    version = node["adapter_version"]
    if type(version) is not str:
        raise ContractError("adapter_version must be text", DOCUMENT_INVALID)
    if _ADAPTER_VERSION.fullmatch(version) is None:
        raise ContractError("adapter_version must be canonical bounded major.minor.patch")


def _validate_declaration_values(node: dict[str, Any]) -> None:
    for field in ("approval_ref", "grant_ref"):
        if node[field] is not None:
            validate_record_ref(node[field], field)
    validate_scope_values(node["scope_ref"])
    for field in ("attempt_ref", "session_ref", "source_snapshot_ref"):
        validate_entity_ref(node[field], field)
    _validate_artifact_set_values(node["input_artifact_refs"], "input_artifact_refs")
    _validate_artifact_set_values(node["output_artifact_refs"], "output_artifact_refs")
    _validate_event_range_values(node["event_range"])
    validate_reason_codes(node["reason_codes"], "reason_codes")


def _validate_artifact_set_values(values: Any, label: str) -> None:
    if type(values) is not list:
        raise ContractError(f"{label} must be an array", DOCUMENT_INVALID)
    if len(values) > MAX_RECEIPT_ARTIFACTS:
        raise ContractError(
            f"{label} must be an array of at most {MAX_RECEIPT_ARTIFACTS} items")
    previous = ""
    for artifact in values:
        _validate_artifact_values(artifact)
        if artifact["logical_id"] <= previous:
            raise ContractError(f"{label} must be strictly sorted by logical_id")
        previous = artifact["logical_id"]


def _validate_declaration_references(node: dict[str, Any]) -> None:
    for field, expected in _RECORD_ROLES.items():
        reference = node[field]
        if reference is not None and reference["record_type"] != expected:
            raise ContractError(
                f"{field} must declare record_type {expected}",
                REFERENCE_MISMATCH)
    for field in ("input_artifact_refs", "output_artifact_refs"):
        for artifact in node[field]:
            _validate_artifact_references(artifact)
            if (artifact["source_snapshot_ref"]["entity_id"] !=
                    node["source_snapshot_ref"]["entity_id"]):
                raise ContractError(
                    "execution artifact source snapshot must match receipt",
                    REFERENCE_MISMATCH)
    event_range = node["event_range"]
    if event_range is not None and event_range["aggregate_ref"] != node["attempt_ref"]:
        raise ContractError("event_range aggregate must equal attempt_ref",
                            REFERENCE_MISMATCH)


def _validate_relations(node: dict[str, Any]) -> None:
    elapsed = node["ended_at_unix_ms"] - node["started_at_unix_ms"]
    if not 0 <= elapsed <= MAX_EXECUTION_ELAPSED_MS:
        raise ContractError("execution wall interval is invalid or exceeds maximum",
                            RELATION_MISMATCH)
    if node["observed_usage"]["elapsed_ms"] != elapsed:
        raise ContractError(
            "observed_usage.elapsed_ms must equal execution wall interval",
            RELATION_MISMATCH)
    for artifact in node["input_artifact_refs"]:
        _validate_artifact_relations(artifact)
        _validate_artifact_relation(artifact, node, False)
    for artifact in node["output_artifact_refs"]:
        _validate_artifact_relations(artifact)
        _validate_artifact_relation(artifact, node, True)
    _validate_event_range_relation(node["event_range"])
    if (node["terminal_state"] == "completed") != (len(node["reason_codes"]) == 0):
        raise ContractError(
            "completed requires no reasons; other terminals require reasons",
            RELATION_MISMATCH)


def _validate_artifact_relation(
        artifact: dict[str, Any], receipt: dict[str, Any], output: bool) -> None:
    if not output and artifact["created_at_unix_ms"] > receipt["started_at_unix_ms"]:
        raise ContractError("input artifact cannot postdate execution start",
                            RELATION_MISMATCH)
    if output and (artifact["producer_attempt_id"] != receipt["attempt_ref"]["entity_id"] or
                   not receipt["started_at_unix_ms"] <= artifact["created_at_unix_ms"] <=
                   receipt["ended_at_unix_ms"]):
        raise ContractError(
            "output artifact must be produced by the attempt during execution",
            RELATION_MISMATCH)


def _validate_event_range_values(value: Any) -> None:
    if value is None:
        return
    node = exact_object(value, EVENT_RANGE_FIELDS, "event_range")
    validate_entity_ref(node["aggregate_ref"], "event_range.aggregate_ref")
    validate_typed_id(node["first_event_id"], "evt", "event_range.first_event_id")
    validate_typed_id(node["last_event_id"], "evt", "event_range.last_event_id")
    validate_i64(node["first_sequence"], "first_sequence", 1)
    validate_i64(node["last_sequence"], "last_sequence", 1)


def _validate_event_range_relation(value: Any) -> None:
    if value is None:
        return
    if value["last_sequence"] < value["first_sequence"]:
        raise ContractError("event_range sequence interval is reversed",
                            RELATION_MISMATCH)
    if (
            (value["first_sequence"] == value["last_sequence"]) !=
            (value["first_event_id"] == value["last_event_id"])):
        raise ContractError(
            "event_range identity and sequence cardinality disagree",
            RELATION_MISMATCH)


def validate_reason_codes(values: Any, label: str) -> None:
    """Validate one exact sorted set of bounded reason tokens."""
    if type(values) is not list:
        raise ContractError(f"{label} must be an array", DOCUMENT_INVALID)
    if len(values) > MAX_REASON_CODES:
        raise ContractError(
            f"{label} must be an array of at most {MAX_REASON_CODES} items")
    previous = ""
    for value in values:
        if type(value) is not str:
            raise ContractError(f"{label} entries must be text", DOCUMENT_INVALID)
        if len(value) > 64 or not lower_token(value) or value <= previous:
            raise ContractError(f"{label} must be strictly sorted unique lowercase tokens")
        previous = value


def _unix_ms(value: Any, label: str) -> int:
    return validate_i64(value, label, 1, MAX_UNIX_MILLISECONDS)

File: forge-runtime/crates/domain/src/platform_core_contract/execution_receipt.rs
use super::{
    ActorType, ArtifactRef, AttemptState, CANONICALIZATION, EntityRef, EntityType, EventRange,
    ExecutionReceipt, ExecutorDescriptor, MAX_EXECUTION_ELAPSED_MS, MAX_OBSERVED_COUNT,
    MAX_OBSERVED_QUANTITY, MAX_REASON_CODES, MAX_RECEIPT_ARTIFACTS, MAX_RECEIPT_BYTES,
    ObservedUsage, PlatformCoreContractError, RECEIPT_VERSION, RejectionCode, artifact, identity,
    recode, references, reject, wire,
};

const APPROVAL_RECORD_TYPE: &str = "forge.control.approval_record";
const GRANT_RECORD_TYPE: &str = "forge.control.capability_grant";

/// Validates supplied execution observations without resolving authority or advancing state.
///
/// # Errors
/// Returns a stable coded error for invalid declarations, references, state, or relations.
pub fn validate_execution_receipt(
    value: &ExecutionReceipt,
) -> Result<(), PlatformCoreContractError> {
    validate_header(value)?;
    validate_time_values(value)?;
    validate_executor(&value.executor)?;
    validate_declaration_values(value)?;
    validate_scope_bindings(value)?;
    validate_declaration_references(value)?;
    validate_terminal_state(value)?;
    validate_relations(value)?;
    wire::canonical_typed(value, MAX_RECEIPT_BYTES).map(|_| ())
}

fn validate_header(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    if value.canonicalization != CANONICALIZATION
        || value.execution_receipt_version != RECEIPT_VERSION
    {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "execution receipt version or canonicalization is unsupported",
        ));
    }
    identity::validate_typed_id(&value.receipt_id, "rcp", "receipt_id")
}

fn validate_scope_bindings(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    references::validate_scope_references(&value.scope_ref)?;
    let Some(snapshot_id) = value.scope_ref.project_snapshot_id.as_deref() else {
        return Err(scope_error());
    };
    let Some(attempt_id) = value.scope_ref.attempt_id.as_deref() else {
        return Err(scope_error());
    };
    let Some(session_id) = value.scope_ref.session_id.as_deref() else {
        return Err(scope_error());
    };
    if value.scope_ref.turn_id.is_some() || value.scope_ref.action_id.is_some() {
        return Err(scope_error());
    }
    validate_exact_ref(
        &value.attempt_ref,
        &EntityType::Attempt,
        attempt_id,
        "attempt_ref",
    )?;
    validate_exact_ref(
        &value.session_ref,
        &EntityType::Session,
        session_id,
        "session_ref",
    )?;
    validate_exact_ref(
        &value.source_snapshot_ref,
        &EntityType::ProjectSnapshot,
        snapshot_id,
        "source_snapshot_ref",
    )
}

fn scope_error() -> PlatformCoreContractError {
    reject(
        RejectionCode::ReferenceMismatch,
        "execution receipt requires exact session-level snapshot scope",
    )
}

fn validate_exact_ref(
    value: &EntityRef,
    entity_type: &EntityType,
    identifier: &str,
    label: &str,
) -> Result<(), PlatformCoreContractError> {
    if &value.entity_type == entity_type && value.entity_id == identifier {
        Ok(())
    } else {
        Err(reject(
            RejectionCode::ReferenceMismatch,
            format!("{label} must exactly match scope_ref"),
        ))
    }
}

fn validate_time_values(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    validate_unix_ms(value.started_at_unix_ms, "started_at_unix_ms")?;
    validate_unix_ms(value.ended_at_unix_ms, "ended_at_unix_ms")?;
    validate_observed_usage(&value.observed_usage)
}

fn validate_terminal_state(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    if !matches!(
        &value.terminal_state,
        AttemptState::Interrupted
            | AttemptState::Completed
            | AttemptState::Failed
            | AttemptState::Uncertain
    ) {
        return Err(reject(
            RejectionCode::StateInvalid,
            "terminal_state is not an Attempt terminal state",
        ));
    }
    Ok(())
}

fn validate_executor(value: &ExecutorDescriptor) -> Result<(), PlatformCoreContractError> {
    references::validate_actor_ref(&value.actor_ref)?;
    if !matches!(
        &value.actor_ref.actor_type,
        ActorType::Agent | ActorType::Service | ActorType::System
    ) {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "executor actor_type must be agent, service, or system",
        ));
    }
    wire::validate_schema_name(&value.adapter_id, "executor.adapter_id")?;
    validate_adapter_version(&value.adapter_version)
}

fn validate_adapter_version(value: &str) -> Result<(), PlatformCoreContractError> {
    if !(5..=32).contains(&value.len()) {
        return Err(adapter_version_error());
    }
    let parts: Vec<&str> = value.split('.').collect();
    if parts.len() != 3 || parts.iter().any(|part| !version_part(part)) {
        return Err(adapter_version_error());
    }
    Ok(())
}

fn version_part(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 9
        && (value.len() == 1 || !value.starts_with('0'))
        && value.bytes().all(|byte| byte.is_ascii_digit())
}

fn adapter_version_error() -> PlatformCoreContractError {
    reject(
        RejectionCode::ValueInvalid,
        "adapter_version must be canonical bounded major.minor.patch",
    )
}

fn validate_observed_usage(value: &ObservedUsage) -> Result<(), PlatformCoreContractError> {
    let checks = [
        (
            "cost_usd_micros",
            value.cost_usd_micros,
            MAX_OBSERVED_QUANTITY,
        ),
        ("elapsed_ms", value.elapsed_ms, MAX_EXECUTION_ELAPSED_MS),
        ("input_tokens", value.input_tokens, MAX_OBSERVED_QUANTITY),
        ("model_calls", value.model_calls, MAX_OBSERVED_COUNT),
        ("network_bytes", value.network_bytes, MAX_OBSERVED_QUANTITY),
        ("output_bytes", value.output_bytes, MAX_OBSERVED_QUANTITY),
        ("output_tokens", value.output_tokens, MAX_OBSERVED_QUANTITY),
        ("tool_calls", value.tool_calls, MAX_OBSERVED_COUNT),
    ];
    for (label, number, maximum) in checks {
        if !(0..=maximum).contains(&number) {
            return Err(reject(
                RejectionCode::ValueInvalid,
                format!("observed_usage.{label} must be in 0..{maximum}"),
            ));
        }
    }
    Ok(())
}

fn validate_declaration_values(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    if let Some(reference) = &value.approval_ref {
        references::validate_record_ref(reference, "approval_ref")?;
    }
    if let Some(reference) = &value.grant_ref {
        references::validate_record_ref(reference, "grant_ref")?;
    }
    validate_reference_values(value)?;
    validate_artifact_set_values(&value.input_artifact_refs, "input_artifact_refs")?;
    validate_artifact_set_values(&value.output_artifact_refs, "output_artifact_refs")?;
    validate_event_range_values(value.event_range.as_ref())?;
    validate_reason_codes(&value.reason_codes, "reason_codes")
}

fn validate_reference_values(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    references::validate_scope_values(&value.scope_ref)?;
    references::validate_entity_ref(&value.attempt_ref, "attempt_ref")?;
    references::validate_entity_ref(&value.session_ref, "session_ref")?;
    references::validate_entity_ref(&value.source_snapshot_ref, "source_snapshot_ref")
}

fn validate_artifact_set_values(
    values: &[ArtifactRef],
    label: &str,
) -> Result<(), PlatformCoreContractError> {
    if values.len() > MAX_RECEIPT_ARTIFACTS {
        return Err(reject(
            RejectionCode::ValueInvalid,
            format!("{label} exceeds {MAX_RECEIPT_ARTIFACTS} items"),
        ));
    }
    let mut previous = "";
    for value in values {
        artifact::validate_artifact_values(value)?;
        if value.logical_id.as_str() <= previous {
            return Err(reject(
                RejectionCode::ValueInvalid,
                format!("{label} must be strictly sorted by logical_id"),
            ));
        }
        previous = &value.logical_id;
    }
    Ok(())
}

fn validate_declaration_references(
    value: &ExecutionReceipt,
) -> Result<(), PlatformCoreContractError> {
    validate_record_role(
        value.approval_ref.as_ref(),
        APPROVAL_RECORD_TYPE,
        "approval_ref",
    )?;
    validate_record_role(value.grant_ref.as_ref(), GRANT_RECORD_TYPE, "grant_ref")?;
    for artifact in value
        .input_artifact_refs
        .iter()
        .chain(&value.output_artifact_refs)
    {
        artifact::validate_artifact_references(artifact)?;
        if artifact.source_snapshot_ref.entity_id != value.source_snapshot_ref.entity_id {
            return Err(reject(
                RejectionCode::ReferenceMismatch,
                "execution artifact source snapshot must match receipt",
            ));
        }
    }
    if value
        .event_range
        .as_ref()
        .is_some_and(|event| event.aggregate_ref != value.attempt_ref)
    {
        return Err(reject(
            RejectionCode::ReferenceMismatch,
            "event_range aggregate must equal attempt_ref",
        ));
    }
    Ok(())
}

fn validate_record_role(
    value: Option<&super::RecordRef>,
    expected: &str,
    label: &str,
) -> Result<(), PlatformCoreContractError> {
    if value.is_some_and(|reference| reference.record_type != expected) {
        return Err(reject(
            RejectionCode::ReferenceMismatch,
            format!("{label} must declare record_type {expected}"),
        ));
    }
    Ok(())
}

fn validate_relations(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    validate_execution_interval(value)?;
    let elapsed = value.ended_at_unix_ms - value.started_at_unix_ms;
    if value.observed_usage.elapsed_ms != elapsed {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "observed_usage.elapsed_ms must equal execution wall interval",
        ));
    }
    for artifact in &value.input_artifact_refs {
        artifact::validate_artifact_relations(artifact)?;
        validate_artifact_relation(artifact, value, false)?;
    }
    for artifact in &value.output_artifact_refs {
        artifact::validate_artifact_relations(artifact)?;
        validate_artifact_relation(artifact, value, true)?;
    }
    validate_event_range_relation(value.event_range.as_ref())?;
    validate_terminal_reason_relation(&value.terminal_state, &value.reason_codes)
}

fn validate_execution_interval(value: &ExecutionReceipt) -> Result<(), PlatformCoreContractError> {
    match value.ended_at_unix_ms.checked_sub(value.started_at_unix_ms) {
        Some(duration) if (0..=MAX_EXECUTION_ELAPSED_MS).contains(&duration) => Ok(()),
        _ => Err(reject(
            RejectionCode::RelationMismatch,
            "execution wall interval is invalid or exceeds maximum",
        )),
    }
}

fn validate_artifact_relation(
    value: &ArtifactRef,
    receipt: &ExecutionReceipt,
    output: bool,
) -> Result<(), PlatformCoreContractError> {
    if !output && value.created_at_unix_ms > receipt.started_at_unix_ms {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "input artifact cannot postdate execution start",
        ));
    }
    if output
        && (value.producer_attempt_id != receipt.attempt_ref.entity_id
            || value.created_at_unix_ms < receipt.started_at_unix_ms
            || value.created_at_unix_ms > receipt.ended_at_unix_ms)
    {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "output artifact must be produced by the attempt during execution",
        ));
    }
    Ok(())
}

fn validate_event_range_values(
    value: Option<&EventRange>,
) -> Result<(), PlatformCoreContractError> {
    let Some(value) = value else {
        return Ok(());
    };
    references::validate_entity_ref(&value.aggregate_ref, "event_range.aggregate_ref")?;
    identity::validate_typed_id(&value.first_event_id, "evt", "event_range.first_event_id")?;
    identity::validate_typed_id(&value.last_event_id, "evt", "event_range.last_event_id")?;
    if value.first_sequence < 1 || value.last_sequence < 1 {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "event_range sequences must be positive",
        ));
    }
    Ok(())
}

fn validate_event_range_relation(
    value: Option<&EventRange>,
) -> Result<(), PlatformCoreContractError> {
    if value.is_some_and(|event| event.last_sequence < event.first_sequence) {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "event_range sequence interval is reversed",
        ));
    }
    if value.is_some_and(|event| {
        (event.first_sequence == event.last_sequence)
            != (event.first_event_id == event.last_event_id)
    }) {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "event_range identity and sequence cardinality disagree",
        ));
    }
    Ok(())
}

fn validate_terminal_reason_relation(
    state: &AttemptState,
    values: &[String],
) -> Result<(), PlatformCoreContractError> {
    if (state == &AttemptState::Completed) == values.is_empty() {
        Ok(())
    } else {
        Err(reject(
            RejectionCode::RelationMismatch,
            "completed requires no reasons; other terminals require reasons",
        ))
    }
}

pub(super) fn validate_reason_codes(
    values: &[String],
    label: &str,
) -> Result<(), PlatformCoreContractError> {
    if values.len() > MAX_REASON_CODES {
        return Err(reject(
            RejectionCode::ValueInvalid,
            format!("{label} exceeds {MAX_REASON_CODES} items"),
        ));
    }
    let mut previous = "";
    for value in values {
        if !wire::lower_token(value) || value.len() > 64 || value.as_str() <= previous {
            return Err(reject(
                RejectionCode::ValueInvalid,
                format!("{label} must be strictly sorted unique lowercase tokens"),
            ));
        }
        previous = value;
    }
    Ok(())
}

fn validate_unix_ms(value: i64, label: &str) -> Result<(), PlatformCoreContractError> {
    if (1..=super::MAX_UNIX_MILLISECONDS).contains(&value) {
        Ok(())
    } else {
        Err(recode(
            super::invalid(format!("{label} is outside the supported range")),
            RejectionCode::ValueInvalid,
        ))
    }
}
